src/net.h

```diff
@@ -1442,9 +1458,34 @@ class CConnman
     std::vector<AddedNodeParams> m_added_node_params GUARDED_BY(m_added_nodes_mutex);
 
     mutable Mutex m_added_nodes_mutex;
+
+    /**
+     * List of peers we are connected with.
+     * Before destroying objects from this list it must be ensured that
+     * all socket operations have completed. `m_msgproc` must be informed
+     * that the node is about to be destroyed by calling `m_msgproc->FinalizeNode()`.
+     * `m_msgproc->ProcessMessages()` and `m_msgproc->SendMessages()` must not be
+     * running on that node during the destruction.
+     *
+     * CNode objects keep a reference count that is used to avoid a destruction
+     * while the object is used. All users of elements from `m_nodes` must either
+     * do so while holding `m_nodes_mutex` or by incrementing the reference before
+     * use and decrement it after use. To make sure that a new reference is not added
+     * to a no-referenced CNode just before it is destroyed we use the following
+     * mechanism (without a help from the standard library):
+     * - New references are only allowed to be added to objects in `m_nodes`.
+     * - When an object from `m_nodes` is to be destroyed it is moved from
+     *   `m_nodes` to `m_nodes_disconnected`. Adding new references to objects
+     *   in `m_nodes_disconnected` is not allowed.
+     * - Once an object in `m_nodes_disconnected` is not referenced anymore,
+     *   then it is safe to destroy it.
+     *
+     * So CNode destruction is: wait for no-references, `FinalizeNode()`, delete CNode.
+     */
     std::vector<CNode*> m_nodes GUARDED_BY(m_nodes_mutex);
-    std::list<CNode*> m_nodes_disconnected;
     mutable RecursiveMutex m_nodes_mutex;
+
+    std::list<CNode*> m_nodes_disconnected;
```

Comment:
Perhaps document how we prevent concurrent access to this while you're at it? (We can't use `GUARDED_BY(m_nodes_mutex)` since that results in lock order issues with cs_main)

AIUI, accesses are:

 * `DisconnectNodes()` via `ThreadSocketHandler()` which adds nodes and removes them in normal operation
 * `StopNodes()` via `Stop()` which runs it after `StopThreads()`, which ensures `threadSocketHandler` is joined before returning. `Stop()` is invoked in `~CConnman()` and `init.cpp:Shutdown()`